component/identity/src/main/java/org/gatein/portal/idm/impl/encoder/Argon2PasswordEncoder.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * This file is part of the Meeds project (https://meeds.io/).
+ * Copyright (C) 2023 Meeds Association
+ * contact@meeds.io
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+*/
+package org.gatein.portal.idm.impl.encoder;
+
+import org.apache.commons.codec.binary.Hex;
+import org.bouncycastle.crypto.generators.Argon2BytesGenerator;
+import org.bouncycastle.crypto.params.Argon2Parameters;
+import org.picketlink.idm.api.Attribute;
+import org.picketlink.idm.api.AttributesManager;
+import org.picketlink.idm.impl.credential.DatabaseReadingSaltEncoder;
+
+import java.nio.charset.StandardCharsets;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+
+public class Argon2PasswordEncoder extends DatabaseReadingSaltEncoder {
+
+  private static final String  PASSWORD_SALT_USER_ATTRIBUTE                      = "passwordSalt128";
+
+  private static final int     ARGON2_HASHING_ITERATIONS                         = 2;
+
+  private static final int     ARGON2_MEMORY_LIMIT                               = 32768;
+
+  private static final int     ARGON2_PARALLEL_EXECUTIONS                        = 2;
+
+  private static final int     ARGON2_HASH_LENGTH                                = 32;
+
+  @Override
+  public String encodeCredential(String username, String rawPassword) {
+    Argon2Parameters.Builder builder =
+                                     new Argon2Parameters.Builder(Argon2Parameters.ARGON2_id).withIterations(ARGON2_HASHING_ITERATIONS)
+                                                                                             .withMemoryAsKB(ARGON2_MEMORY_LIMIT)
+                                                                                             .withParallelism(ARGON2_PARALLEL_EXECUTIONS)
+                                                                                             .withSalt(getStoredSalt(username));
+    Argon2BytesGenerator generator = new Argon2BytesGenerator();
+    generator.init(builder.build());
+    byte[] hash = new byte[ARGON2_HASH_LENGTH];
+    generator.generateBytes(rawPassword.getBytes(StandardCharsets.UTF_8), hash, 0, hash.length);
```

Comment:
In the specification, we speak about a symetrical encryption at the end of the encode process, using CodecInitializer.

Could you add it ?

Reply:
Done